### ISSUE TITLE
Regression: firedrake-adjoint and VertexOnlyMesh hessian evaluation failure

### ISSUE THREAD

**Issue by ReubenHill** [firedrakeproject/firedrake#2685]
The test
`tests/vertexonly/test_poisson_inverse_conductivity.py`
fails if I increase the number of points to, say `num_points = 1024`.
```
tests/vertexonly/test_poisson_inverse_conductivity.py Fatal Python error: Segmentation fault

Current thread 0x0000000108f4c600 (most recent call first):
  File "/Users/rwh10/firedrake/src/firedrake/firedrake/linear_solver.py", line 159 in solve
  File "/Users/rwh10/firedrake/src/firedrake/firedrake/solving.py", line 244 in _la_solve
  File "/Users/rwh10/firedrake/src/firedrake/firedrake/solving.py", line 133 in solve
  File "/Users/rwh10/firedrake/src/firedrake/firedrake/adjoint/solving.py", line 53 in wrapper
  File "/Users/rwh10/firedrake/src/pyadjoint/dolfin_adjoint_common/blocks/solving.py", line 173 in _assemble_and_solve_adj_eq
  File "/Users/rwh10/firedrake/src/pyadjoint/dolfin_adjoint_common/blocks/solving.py", line 148 in prepare_evaluate_adj
  File "/Users/rwh10/firedrake/src/pyadjoint/pyadjoint/block.py", line 128 in evaluate_adj
  File "/Users/rwh10/firedrake/src/pyadjoint/pyadjoint/tape.py", line 69 in wrapper
  File "/Users/rwh10/firedrake/src/pyadjoint/pyadjoint/tape.py", line 192 in evaluate_adj
  File "/Users/rwh10/firedrake/src/pyadjoint/pyadjoint/drivers.py", line 29 in compute_gradient
  File "/Users/rwh10/firedrake/src/pyadjoint/pyadjoint/reduced_functional.py", line 64 in derivative
  File "/Users/rwh10/firedrake/src/pyadjoint/pyadjoint/reduced_functional_numpy.py", line 70 in derivative
  File "/Users/rwh10/firedrake/src/pyadjoint/pyadjoint/tape.py", line 69 in wrapper
  File "/Users/rwh10/firedrake/src/pyadjoint/pyadjoint/reduced_functional_numpy.py", line 87 in hessian
  File "/Users/rwh10/firedrake/src/pyadjoint/pyadjoint/tape.py", line 69 in wrapper
  File "/Users/rwh10/firedrake/lib/python3.10/site-packages/scipy/optimize/_optimize.py", line 1930 in _minimize_newtoncg
  File "/Users/rwh10/firedrake/lib/python3.10/site-packages/scipy/optimize/_minimize.py", line 696 in minimize
  File "/Users/rwh10/firedrake/src/pyadjoint/pyadjoint/optimization/optimization.py", line 135 in minimize_scipy_generic
  File "/Users/rwh10/firedrake/src/pyadjoint/pyadjoint/optimization/optimization.py", line 252 in minimize
  File "/Users/rwh10/firedrake/src/pyadjoint/pyadjoint/tape.py", line 69 in wrapper
  File "/Users/rwh10/firedrake/src/firedrake/tests/vertexonly/test_poisson_inverse_conductivity.py", line 94 in test_poisson_inverse_conductivity
```
This is fixed by switching line 94
```
minimize(Ĵ, method='Newton-CG', options={'disp': True})
```
to
```
minimize(Ĵ, options={'disp': True})
```
which suggests the issue is in the adjoint solve that happens in the hessian action evaluation.

EDIT:
`firedrake-status` reports
```
Firedrake Configuration:
    package_manager: True
    minimal_petsc: False
    mpicc: None
    mpicxx: None
    mpif90: None
    mpiexec: None
    disable_ssh: False
    honour_petsc_dir: False
    with_parmetis: False
    slepc: False
    packages: []
    honour_pythonpath: False
    opencascade: False
    tinyasm: False
    petsc_int_type: int32
    cache_dir: /Users/rwh10/firedrake/.cache
    complex: False
    remove_build_files: False
    with_blas: /usr/local/opt/openblas
Additions:
    None
Environment:
    PYTHONPATH: None
    PETSC_ARCH: None
    PETSC_DIR: None
Status of components:
---------------------------------------------------------------------------
|Package             |Branch                        |Revision  |Modified  |
---------------------------------------------------------------------------
|COFFEE              |master                        |70c1e66   |False     |
|FInAT               |master                        |3f28113   |False     |
|PyOP2               |master                        |7aca5e50  |True      |
|fiat                |master                        |43bc840   |False     |
|firedrake           |master                        |2209aaeb  |True      |
|h5py                |firedrake                     |78531f08  |False     |
|libspatialindex     |master                        |4768bf3   |True      |
|libsupermesh        |master                        |69012e5   |False     |
|loopy               |main                          |3988272b  |False     |
|petsc               |firedrake                     |7964b45c3c|False     |
|pyadjoint           |master                        |4742834   |False     |
|tsfc                |master                        |ceebdd5   |False     |
|ufl                 |master                        |0c592ec5  |False     |
---------------------------------------------------------------------------
```
Platform is intel macos Monterey (12.6).

**Comment by wence-** [comment]
Can you get a backtrace?

**Comment by ReubenHill** [comment]
> Can you get a backtrace?

@wence- I'm struggling for this example - I need to do more than just `export PYOP2_DEBUG=1` right?

**Comment by ReubenHill** [comment]
Todo: see if this occurs on other platforms

**Comment by wence-** [comment]
> @wence- I'm struggling for this example - I need to do more than just `export PYOP2_DEBUG=1` right?

`gdb --args python script-that-provokes-failure.py`

**Comment by ReubenHill** [comment]
from `lldb -- python failing_test.py` with the test refactored into a script
```
    frame #0: 0x00007ff80de7b00e libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007ff80deb11ff libsystem_pthread.dylib`pthread_kill + 263
    frame #2: 0x00007ff80ddfcd24 libsystem_c.dylib`abort + 123
    frame #3: 0x00007ff80dcda357 libsystem_malloc.dylib`malloc_vreport + 551
    frame #4: 0x00007ff80dcee308 libsystem_malloc.dylib`malloc_zone_error + 178
    frame #5: 0x00007ff80dccd0e8 libsystem_malloc.dylib`nanov2_allocate_from_block + 582
    frame #6: 0x00007ff80dccc677 libsystem_malloc.dylib`nanov2_allocate + 130
    frame #7: 0x00007ff80dccc5a4 libsystem_malloc.dylib`nanov2_malloc + 56
    frame #8: 0x00007ff80dce8abb libsystem_malloc.dylib`_malloc_zone_malloc + 125
    frame #9: 0x00007ff80dd7c6d6 libsystem_c.dylib`strdup + 32
  * frame #10: 0x0000000104c10f96 libpmpi.12.dylib`MPIR_Err_create_code_valist [inlined] vsnprintf_mpi(str="", maxlen=256, fmt_orig="MPI_Comm_free(comm=%p) failed", list=<unavailable>) at errutil.c:1519:11 [opt]
    frame #11: 0x0000000104c10f8e libpmpi.12.dylib`MPIR_Err_create_code_valist(lastcode=813198341, fatal=0, fcname=<unavailable>, line=87, error_class=<unavailable>, generic_msg="", specific_msg="**mpi_comm_free %p", Argp=0x00007ff7bfef8680) at errutil.c:984:21 [opt]
    frame #12: 0x0000000104c10a84 libpmpi.12.dylib`MPIR_Err_create_code(lastcode=<unavailable>, fatal=<unavailable>, fcname=<unavailable>, line=<unavailable>, error_class=<unavailable>, generic_msg=<unavailable>, specific_msg="**mpi_comm_free %p") at errutil.c:854:10 [opt]
    frame #13: 0x0000000103b5d52d libmpi.12.dylib`MPI_Comm_free [inlined] internal_Comm_free(comm=0x00007ff7bfef8e78) at comm_free.c:87:17 [opt]
    frame #14: 0x0000000103b5d201 libmpi.12.dylib`MPI_Comm_free(comm=0x00007ff7bfef8e78) at comm_free.c:156:12 [opt]
    frame #15: 0x00000001039b3941 libmpifort.12.dylib`mpi_comm_free_(comm=0x00007ff7bfef8e78, ierr=0x00007ff7bfef8e30) at fortran_binding.c:2394:13 [opt]
    frame #16: 0x0000000106e10100 libpetsc.3.018.2.dylib`dmumps_fac_driver_ at dfac_driver.F:1030:72
    frame #17: 0x0000000106e8f42e libpetsc.3.018.2.dylib`dmumps_ at dmumps_driver.F:1916:72
    frame #18: 0x0000000106e92c74 libpetsc.3.018.2.dylib`dmumps_f77_ at dmumps_f77.F:289:72
    frame #19: 0x0000000106e8a806 libpetsc.3.018.2.dylib`dmumps_c(mumps_par=0x000000012ff87c00) at mumps_c.c:485:5 [opt]
    frame #20: 0x000000010629c432 libpetsc.3.018.2.dylib`MatFactorNumeric_MUMPS(F=0x000000012ff87000, A=0x000000012feb7e00, info=<unavailable>) at mumps.c:1681:3 [opt]
    frame #21: 0x0000000106501aa2 libpetsc.3.018.2.dylib`MatLUFactorNumeric(fact=0x000000012ff87000, mat=0x000000012feb7e00, info=0x0000000137816d78) at matrix.c:3206:3 [opt]
    frame #22: 0x0000000106a0d269 libpetsc.3.018.2.dylib`PCSetUp_LU(pc=0x00000001378169f0) at lu.c:120:5 [opt]
    frame #23: 0x0000000106a574ca libpetsc.3.018.2.dylib`PCSetUp(pc=0x00000001378169f0) at precon.c:994:5 [opt]
    frame #24: 0x0000000106af25bc libpetsc.3.018.2.dylib`KSPSetUp(ksp=0x000000012feb7000) at itfunc.c:405:3 [opt]
    frame #25: 0x0000000106af4478 libpetsc.3.018.2.dylib`KSPSolve_Private(ksp=0x000000012feb7000, b=0x000000012ff8ca00, x=<unavailable>) at itfunc.c:824:3 [opt]
    frame #26: 0x0000000106af429e libpetsc.3.018.2.dylib`KSPSolve(ksp=<unavailable>, b=<unavailable>, x=<unavailable>) at itfunc.c:1070:3 [opt]
    frame #27: 0x0000000103fea95b PETSc.cpython-310-darwin.so`__pyx_pw_8petsc4py_5PETSc_3KSP_101solve [inlined] __pyx_pf_8petsc4py_5PETSc_3KSP_100solve(__pyx_v_self=<unavailable>, __pyx_v_b=0x0000000000000000, __pyx_v_x=0x0000000000000000) at petsc4py.PETSc.c:219334:47 [opt]
    frame #28: 0x0000000103fea94c PETSc.cpython-310-darwin.so`__pyx_pw_8petsc4py_5PETSc_3KSP_101solve(__pyx_v_self=<unavailable>, __pyx_args=0x000000012ecc86c0, __pyx_kwds=<unavailable>) at petsc4py.PETSc.c:219256:13 [opt]
    frame #29: 0x00000001005a3b85 Python`method_vectorcall_VARARGS_KEYWORDS + 179
    frame #30: 0x000000010069144c Python`call_function + 172
    frame #31: 0x000000010068858b Python`_PyEval_EvalFrameDefault + 24074
    frame #32: 0x00000001006814b5 Python`_PyEval_Vector + 342
    frame #33: 0x00000001040e5515 PETSc.cpython-310-darwin.so`__pyx_pw_8petsc4py_5PETSc_3Log_14EventDecorator_9decorator_1wrapped_func [inlined] __Pyx_PyObject_Call(func=0x0000000124ca2b00, arg=0x00000001344d1b80, kw=0x000000012ec35c40) at petsc4py.PETSc.c:408530:14 [opt]
    frame #34: 0x00000001040e54e0 PETSc.cpython-310-darwin.so`__pyx_pw_8petsc4py_5PETSc_3Log_14EventDecorator_9decorator_1wrapped_func at petsc4py.PETSc.c:77660:23 [opt]
    frame #35: 0x00000001040e4efa PETSc.cpython-310-darwin.so`__pyx_pw_8petsc4py_5PETSc_3Log_14EventDecorator_9decorator_1wrapped_func(__pyx_self=<unavailable>, __pyx_args=<unavailable>, __pyx_kwds=<unavailable>) at petsc4py.PETSc.c:77458:13 [opt]
    frame #36: 0x0000000100596e3f Python`_PyObject_MakeTpCall + 132
    frame #37: 0x000000010059a4e8 Python`method_vectorcall + 688
    frame #38: 0x000000010069144c Python`call_function + 172
    frame #39: 0x0000000100689bb2 Python`_PyEval_EvalFrameDefault + 29745
    frame #40: 0x00000001006814b5 Python`_PyEval_Vector + 342
    frame #41: 0x00000001006833ee Python`_PyEval_EvalFrameDefault + 3181
    frame #42: 0x00000001006814b5 Python`_PyEval_Vector + 342
    frame #43: 0x00000001006833ee Python`_PyEval_EvalFrameDefault + 3181
    frame #44: 0x00000001006814b5 Python`_PyEval_Vector + 342
    frame #45: 0x00000001040e5515 PETSc.cpython-310-darwin.so`__pyx_pw_8petsc4py_5PETSc_3Log_14EventDecorator_9decorator_1wrapped_func [inlined] __Pyx_PyObject_Call(func=0x0000000124ce4dc0, arg=0x0000000134452d40, kw=0x00000001344da180) at petsc4py.PETSc.c:408530:14 [opt]
    frame #46: 0x00000001040e54e0 PETSc.cpython-310-darwin.so`__pyx_pw_8petsc4py_5PETSc_3Log_14EventDecorator_9decorator_1wrapped_func at petsc4py.PETSc.c:77660:23 [opt]
    frame #47: 0x00000001040e4efa PETSc.cpython-310-darwin.so`__pyx_pw_8petsc4py_5PETSc_3Log_14EventDecorator_9decorator_1wrapped_func(__pyx_self=<unavailable>, __pyx_args=<unavailable>, __pyx_kwds=<unavailable>) at petsc4py.PETSc.c:77458:13 [opt]
```

**Comment by ReubenHill** [comment]
This is fixed by #2674
